aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,234 @@
+    use aptos_std::error;
+    friend aptos_framework::genesis;
+    friend aptos_framework::transaction_validation;
+
+    const NUM_BUCKETS: u64 = 50000;
+
+    // Max expiration time for orderless txns.
+    // We will advertise to the community that max expiration time for orderless txns is 60 seconds.
+    // Adding a 5 second slack here as the client's time and the blockchain's time may drift.
+    const MAX_EXPIRATION_TIME: u64 = 65;
+
+    // Each time we check if an (address, nonce) pair can be inserted into nonce history,
+    // we try to initially garbage collect expired nonces in the bucket. This is a limit on the number of nonces
+    // we try to garbage collect in a single call.
+    const MAX_ENTRIES_GARBAGE_COLLECTED_PER_CALL: u64 = 5;
+
+    // Nonce history resource does not exist.
+    const E_NONCE_HISTORY_DOES_NOT_EXIST: u64 = 1001;
+
+    // An orderless transaction is a transaction that doesn't have a sequence number.
+    // Orderless transactions instead contain a nonce to prevent replay attacks.
+    // If the incoming transaction has the same (address, nonce) pair as a previous unexpired transaction, it is rejected.
+    // The nonce history is used to store the list of (address, nonce, txn expiration time) values of all unexpired transactions.
+    // The nonce history is used in the transaction validation process to check if the incoming transaction is valid.
+    struct NonceHistory has key {
+        // Key = sip_hash(NonceKey) % NUM_BUCKETS
+        // Value = Bucket
+        nonce_table: Table<u64, Bucket>,
+        // Used to facilitate prefill the nonce_table with empty buckets
+        // one by one using `add_nonce_bucket` method.
+        // This is the next_key to prefill with an empty bucket
+        next_key: u64,
+    }
+
+    // The bucket stores (address, nonce, txn expiration time) tuples.
+    // All the entries in the bucket contain the same hash(address, nonce) % NUM_BUCKETS.
+    // The first big ordered map in the bucket stores (expiration time, address, nonce) -> true.
+    // The second big ordered map in the bucket stores (address, nonce) -> expiration time.
+    // Both the maps store the same data, just in a different format.
+    // As the key in the first big ordered map starts with expiration time, it's easy to figure out which
+    // entries have expired at the current time. The first big ordered map helps with easy garbage collection.
+    // The second big ordered map helps with checking if the given (address, nonce) pair exists in the bucket.
+    // An (address, nonce) pair is guaranteed to be unique in both the big ordered maps. Two transactions with
+    // the same (address, nonce) pair cannot be stored at the same time.
+    struct Bucket has store {
+        // The first big ordered map in the bucket stores (expiration time, address, nonce) -> true.
+        nonces_ordered_by_exp_time: BigOrderedMap<NonceKeyWithExpTime, bool>,
```

Comment:
have you tried `smart_vector<u64, smart_vector<nouce_key>>`?

Reply:
I tried using vector. The performance of vector gets bad when the size of the bucket increases. Big ordered map has O(log n) complexity.